doc/developer/design/20220226_multiwriter_persist.md

```diff
@@ -0,0 +1,205 @@
+## Overview
+
+This document describes the design of a persistent collection implementation
+with the following properties:
+
+* Supports any number of writers that can come and go with zero coordination
+* Supports any number of readers that can read with zero coordination
+* Background compaction and garbage collection of no longer needed batches with
+  zero coordination with readers or writers
+
+The design is based on Differential Dataflow's [Spine][spine] data structure.
+
+## Collection structure
+
+A persistent collection is made up of consolidated batches of updates. Each
+batch is described by a `lower` and `upper` frontier and contains all the
+updates to the persistent collection that happened at times beyond `lower` and
+not beyond `upper`. Batches are immutable and are never modified after their
+creation. When they are no longer needed they are garbage collected.
+
+The batches are organized in two separate areas, a **versioned** "merged area"
+and a "pending area" that is sharded by writer id.
```

Comment:
I realized that `shard` might be misleading there. I don't mean that each writer is responsible for a different subset of the data. All writers write the same data and the pending area is just subdivided to give each writer a dedicated place to put their batches

Reply:
If there is no good alternative to `shard` that makes this point clear, we should add a sentence about this to the main text (e.g. once this is merged, the comments are not visible directly).

Reply:
I changed the description to not use the term shard